internal/distro/distro.go

```diff
@@ -275,3 +275,55 @@ func ExportsFallback() []string {
 func PayloadPackageSets() []string {
 	return []string{}
 }
+
+func MakePackageSetChains(t ImageType, packageSets map[string]rpmmd.PackageSet, repos []rpmmd.RepoConfig) map[string][]rpmmd.PackageSet {
+	allSetNames := make([]string, len(packageSets))
+	idx := 0
+	for setName := range packageSets {
+		allSetNames[idx] = setName
+		idx++
+	}
+
+	// map repository PackageSets to the list of repo configs
+	packageSetsRepos := make(map[string][]rpmmd.RepoConfig)
+	for idx := range repos {
+		repo := repos[idx]
+		if len(repo.PackageSets) == 0 {
+			// repos that don't specify package sets get used everywhere
+			repo.PackageSets = allSetNames
+		}
+		for _, name := range repo.PackageSets {
+			psRepos := packageSetsRepos[name]
+			psRepos = append(psRepos, repo)
+			packageSetsRepos[name] = psRepos
+		}
+	}
+
+	chainedSets := make(map[string][]rpmmd.PackageSet)
+	addedSets := make(map[string]bool)
+	// first collect package sets that are part of a chain
+	for specName, setNames := range t.PackageSetsChains() {
+		pkgSets := make([]rpmmd.PackageSet, len(setNames))
+
+		// add package-set-specific repositories to each set if one is defined
+		for idx, pkgSetName := range setNames {
+			pkgSets[idx] = packageSets[pkgSetName]
```

Comment:
Looking at this code again and thinking about the removed unit test that previously ensured that package set names used in package set chains are defined for the image type, I think that it would be worth to change this line to check that the `pkgSetName` exists in `packageSets` and return `error` from the function if it does not. Otherwise a default type value is returned and it would hide a programming error.

It would be even justified to `panic` here or in the caller function in case an error is returned.

Then we could replace the removed unit test by simply calling `PackageSets()` on every image type with `nil` as repos and it would produce error / `panic` in case the package set chain uses a package set name not defined for the image type.

Reply:
Good idea.  I brought back the test and added it to `internal/distro/distro_test.go` so it's run with the unit tests and I also added the `panic()` for when a package set defined in a chain doesn't exist in the package sets for the image type.

The test itself compares the keys between `PackageSets()` and `ChainedPackageSets()` and the function itself takes care of panicking.

You're right, it's all much better this way.